### PR TITLE
Symbolic sizes

### PR DESCRIPTION
**Pull request by S1eGa**

### CODE REVIEW

File: lib/Expr/ExprPPrinter.cpp [hunk 1]
```diff
@@ -520,8 +520,11 @@ void ExprPPrinter::printQuery(llvm::raw_ostream &os,
...
       } else {
+        ref<ConstantExpr> sizeConstantExpr = cast<ConstantExpr>(
+            constraints.getConcretization().evaluate(A->size));
         PC << "[";
-        for (unsigned i = 0, e = A->size; i != e; ++i) {
+        for (unsigned i = 0, e = sizeConstantExpr->getZExtValue(); i != e;
+             ++i) {
           if (i)
             PC << " ";
           PC << A->constantValues[i];
```
Comment (S1eGa) [inline lib/Expr/ExprPPrinter.cpp:530]: What if source is constantWithSymbolicSizeSource?

File: lib/Core/Executor.cpp [hunk 2]
```diff
@@ -4334,12 +4543,151 @@ void Executor::resolveExact(ExecutionState &state,
...
+  solver->setTimeout(time::Span());
+
+  if (!success) {
+    return nullptr;
+  }
+
+  /* We can simply query the solver to get value of size, but
+  to optimize the number of queries we will ask it one time with assignment
+  for symcretes. */
+  bool checkSize;
```
Comment (S1eGa) [inline lib/Core/Executor.cpp:4614]: Do we need thus check?

File: lib/Core/AddressSpace.cpp [hunk 3]
```diff
@@ -92,119 +93,167 @@ bool AddressSpace::resolveOne(const ref<ConstantExpr> &addr, KType *objectType,
...
+      return false;
+    if (mayBeTrue) {
+      result = oi->first->id;
+      success = true;
+      return true;
+    } else {
       bool mustBeTrue;
       if (!solver->mustBeTrue(state.constraints,
-                              UltExpr::create(address, mo->getBaseExpr()),
+                              UgeExpr::create(address, mo->getBaseExpr()),
```
Comment (S1eGa) [inline lib/Core/AddressSpace.cpp:215]: If addresses can recompute, then we do not need this check.

File: lib/Core/Executor.cpp [hunk 4]
```diff
@@ -4684,36 +4997,66 @@ IDType Executor::lazyInitializeObject(ExecutionState &state,
...
+          cast<ConstantExpr>(assignment.evaluate(Expr::createTempRead(
+                                 bind.first, Context::get().getPointerWidth())))
+              ->getZExtValue();
+      MemoryObject *newMO =
+          addressManager->allocateMemoryObject(dependentAddressArray, newSize);
+
+
+      ObjectPair oldOp = state.addressSpace.findObject(newMO->id);
+      const MemoryObject *oldMO = oldOp.first;
+      const ObjectState *oldOS = oldOp.second;
```
Comment (misonijnik) [inline lib/Core/Executor.cpp:5027]: Should be wrapped in ref<ObjectState>

File: lib/Solver/SolverImpl.cpp [hunk 5]
```diff
@@ -53,6 +54,130 @@ bool SolverImpl::computeValidityCore(const Query &query,
...
+      }
+
+      if (mustBeGreater) {
+        left = middle;
+      } else {
+        right = middle;
+      }
+    }
+    left = ShlExpr::create(ConstantExpr::create(1, left->getWidth()), left);
+    right = ShlExpr::create(ConstantExpr::create(1, right->getWidth()), right);
```
Comment (misonijnik) [inline lib/Solver/SolverImpl.cpp:134]: May overflow

File: lib/Solver/SolverImpl.cpp [hunk 6]
```diff
@@ -53,6 +54,130 @@ bool SolverImpl::computeValidityCore(const Query &query,
...
+    } while (mustBeTrue);
+  }
+
+  // Binary search the least value for expr from the given query
+  while (cast<ConstantExpr>(
+             UltExpr::create(AddExpr::create(left, ConstantExpr::create(
+                                                       1, left->getWidth())),
+                             right))
+             ->isTrue()) {
+    ref<ConstantExpr> middle = cast<ConstantExpr>(
```
Comment (misonijnik) [inline lib/Solver/SolverImpl.cpp:159]: may overflow

File: lib/Core/AddressSpace.cpp [hunk 7]
```diff
@@ -92,119 +93,150 @@ bool AddressSpace::resolveOne(const ref<ConstantExpr> &addr, KType *objectType,
...
+  }
 
-    // search forwards
-    for (oi=start; oi!=end; ++oi) {
-      const auto &mo = oi->first;
-      if (!predicate(mo))
-        continue;
+  // didn't work, now we have to search
+      
+  MemoryMap::iterator oi = objects.upper_bound(&hack);
```
Comment (misonijnik) [inline lib/Core/AddressSpace.cpp:188]: Does not work correctly with ConcretizingSolver

**Pull request merged**